server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java

```diff
@@ -142,6 +147,37 @@ public void testCancelException() {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
+    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+        final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+        final var threadCount = 4;
+        final var running = new AtomicBoolean(true);
+        final var startBarrier = new CyclicBarrier(threadCount + 1);
+        final var futures = new ArrayList<Future<?>>();
+        try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
+            assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
+            // N threads competing to complete the future
+            for (int i = 0; i < threadCount; i++) {
+                futures.add(threadPool.executor(executorName).submit(() -> {
+                    safeAwait(startBarrier);
+                    while (running.get()) {
+                        futureReference.get().onResponse(null);
+                    }
+                }, null));
+            }
+            // continually create new futures to complete
+            safeAwait(startBarrier);
+            long endTime = System.currentTimeMillis() + 20;
+            while (System.currentTimeMillis() < endTime) {
+                futureReference.set(new PlainActionFuture<>());
+                safeSleep(1);
+            }
+            running.set(false);
+            futures.forEach(ESTestCase::safeGet);
```

Comment:
No real need to wait on these futures, uncaught exceptions will fail the test anyway and shutting down the threadpool will `join()` all the threads.